src/Diagnostics.DataProviders/AscClient.cs

```diff
@@ -0,0 +1,378 @@
+                {
+                    requestMessage.Headers.Add("Authorization", authToken);
+
+                    httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", authToken);
+
+                    var response = await httpClient.SendAsync(requestMessage, cancellationToken);
+                    string responseContent = await response.Content.ReadAsStringAsync();
+                    if (response.IsSuccessStatusCode)
+                    {
+                        if (typeof(T).Equals(typeof(string)))
+                        {
+                            return CastTo<T>(responseContent);
+                        }
+                        else
+                        {
+                            T value;
+                            try
+                            {
+                                value = JsonConvert.DeserializeObject<T>(responseContent);
+                            }
+                            catch (Exception serializeException)
+                            {
+                                throw new Exception($" Failed to serialize ASC response to type {typeof(T).ToString()} : Response from ASC ==> {responseContent}", serializeException);
+                            }
+
+                            return value;
+                        }
+                    }
+                    else
+                    {
+                        string responseLogBody = await GetRequestDetailsForLogging(response);
+                        throw new Exception(string.Format("HTTP GET request to ASC failed. AppLens request Id : {0} ==> Details : {1}", requestId, responseLogBody));
+                    }
+                }
+            }
+            catch (TaskCanceledException ex)
+            {
+                if (cancellationToken != default(CancellationToken))
+                {
+                    throw new Exception(string.Format("The HTTP GET request to ASC with query string {0} was cancelled as per the supplied cancellation token. AppLens request Id : {1}", queryString, requestId), ex);
+                }
+                else
+                {
+                    throw;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<T> GetInsightFromBlob<T>(string blobUri, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                httpClient.DefaultRequestHeaders.Remove("Authorization");
+                httpClient.DefaultRequestHeaders.UserAgent.Clear();
+                httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue(new ProductHeaderValue(userAgent)));
+                Uri blobUriObj;
+                if (string.IsNullOrEmpty(blobUri) || !Uri.TryCreate(blobUri, UriKind.RelativeOrAbsolute, out blobUriObj))
+                {
+                    throw new Exception($"Invalid blob uri {blobUri}.");
```

Comment:
Can this be `ArgumentNullException(nameof(blobUri))`?

Reply:
I broke the two checks into individual checks and am throwing a more apt exception for each case now.